src/web/js/output-ui.js

```diff
@@ -685,278 +443,234 @@
+          readOnly:       "nocursor",
+          disableInput:   true,
+          indentUnit:     2,
+          lineWrapping:   false,
+          lineNumbers:    true,
+          viewportMargin: 1,
+          scrollbarStyle: "null"});
+        editor.swapDoc(new CodeMirror.Doc(
+          lines, 
+           position.doc.mode, 
+           position.from.line,
+           position.doc.lineSep));
+        this.container = container;
+        this.position  = position;
+        this.editor    = editor;
+        this.doc       = editor.doc;
+      }
+      return Snippet;
+    }();
+    
+    function renderStackTrace(runtime, documents, srcloc, pyretStack) {
+      function isSrcloc(s) { 
+        return s && runtime.unwrap(runtime.getField(srcloc, "is-srcloc").app(s));
+      }
+      var container = $("<div>").addClass("stacktrace");
+      container.append($("<p>").text("Evaluation in progress when the error occurred (most recent first):"));
+      container.on("mouseover", function () {
+        $("#main").addClass("spotlight");
+      });
+      container.on("mouseleave", function () {
+        $("#main").removeClass("spotlight");
+      });
+      pyretStack.
+        map(runtime.makeSrcloc).
+        filter(isSrcloc).
+        map(function (loc) {
+          if (!documents.has(loc.dict.source)) {
+            return $('<div>').append(drawSrcloc(documents, runtime, loc).css('display', 'block'));
           } else {
-            return null;
+            var position = Position.fromPyretSrcloc(runtime, srcloc, loc, documents);
+            var snippet  = new Snippet(position);
+            var spotlight;
+            snippet.container.addEventListener("mouseover", function() {
+              if(spotlight !== undefined)
+                return;
+              spotlight = position.spotlight();
+            });
+            snippet.container.addEventListener("mouseleave", function() {
+              if(spotlight !== undefined)
+                spotlight.clear();
+              spotlight = undefined;
+            });
+            return $(snippet.container);
           }
+        }).
+        forEach(function(frame) {
+          container.append(frame);
+        });
+      return expandable(container, "program execution trace");
+    }
+    
+    var allHighlightAnchors   = new Map();
+    var allHighlightPositions = new Map();
+    var colorsEmphasized      = new Set();
+    var colorsHighlighted     = new Set();
+    lastHue = (lastHue + goldenAngle)%(Math.PI*2.0);
+    var globalColor = hueToRGB(lastHue);
+    
+    function highlight(color) {
+      if(colorsHighlighted.has(color))
+        return;
+      else {
+        colorsHighlighted.add(color);
+        var anchors   = allHighlightAnchors.get(color);
+        var positions = allHighlightPositions.get(color);
+        for(var i = 0; i < anchors.length; i++) {
+          anchors[i].css('background-color',
+            sessionStorage.getItem("highlight-colorfulness") != "vibrant" 
```

Comment:
Partially addressed by 1d5fc0fe85e954303acd5ca56735208373c5bf0d. This lifts the `sessionStorage` access out of the loop, but not parameterize the functions. I'm a little wary of working to support a feature that doesn't have an immediate or foreseeable need. Parameterizing the functions is straightforward, but manipulating highlight mode with some other mechanism would require a good deal of careful thought. It seems prudent to cross that bridge when we get there.